### PR TITLE
Refactor the store and the module instance.

### PR DESCRIPTION
**Pull request by q82419**
- [x] Move the ownership of the instances from the store manager to the module instance.
- [x] Merge the import object into the module instance.
- [x] Invoke functions by the function instance instead of by function name in executor.
- [x] Refactor the WasmEdge C API.
- [x] Update the tests.
- [x] Add the `Upgrade to 0.10.0` guideline.
- [x] Update the C API document.

### CODE REVIEW

**Comment by apepkuss** [comment]
> Details

@q82419 Thanks for the reminder. I checked the number of the refactored C-APIs, and it would be better to launch a separate PR to migrate rust bindings after merging this PR. Currently, the failed binding-rust workflows can be ignored.

**Comment by apepkuss** [comment]
Is it possible to expose the following two APIs?
- Statistics.clear()
- StoreManager.reset()

In addition, I noticed that in the current design `Executor` provides the APIs for registering `ASTModule` and `ImportObject`. Is it possible to move the registering APIs from `Executor` to `Store`? So that in the scenario, like multiple threading, we can only `move` or `share` a store between threads. I'm not very sure if the idea is appropriate or not in real practice.

**Comment by q82419** [comment]
> Is it possible to expose the following two APIs?
> 
> * Statistics.clear()
> * StoreManager.reset()
> 
> In addition, I noticed that in the current design `Executor` provides the APIs for registering `ASTModule` and `ImportObject`. Is it possible to move the registering APIs from `Executor` to `Store`? So that in the scenario, like multiple threading, we can only `move` or `share` a store between threads. I'm not very sure if the idea is appropriate or not in real practice.

For the statistics: ok, I'll expose later.
For the store manager: in the current PR, the ownership is moved into the module instance, developers should keep and handle the module instance by themselves. The store manager only record the module exporting for linking when instantiation. I think it's not necessary to implement the reset function. By the way, it's better to remove the export info in the store manager when a named module being destroyed. I'll update for this.

It's not appropriate to move the registering APIs from the `Executor` to `Store`. The `Store` means the object to handle the instances for searching and module linking, and the `Executor` is the object to instantiate and register WASM modules.

**Comment by apepkuss** [comment]
> > Is it possible to expose the following two APIs?
> > 
> > * Statistics.clear()
> > * StoreManager.reset()
> > 
> > In addition, I noticed that in the current design `Executor` provides the APIs for registering `ASTModule` and `ImportObject`. Is it possible to move the registering APIs from `Executor` to `Store`? So that in the scenario, like multiple threading, we can only `move` or `share` a store between threads. I'm not very sure if the idea is appropriate or not in real practice.
> 
> For the statistics: ok, I'll expose later. For the store manager: in the current PR, the ownership is moved into the module instance, developers should keep and handle the module instance by themselves. The store manager only record the module exporting for linking when instantiation. I think it's not necessary to implement the reset function. By the way, it's better to remove the export info in the store manager when a named module being destroyed. I'll update for this.
> 
> It's not appropriate to move the registering APIs from the `Executor` to `Store`. The `Store` means the object to handle the instances for searching and module linking, and the `Executor` is the object to instantiate and register WASM modules.

@q82419 Thanks for the explanation.

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

File: include/runtime/instance/module.h [hunk 1]
```diff
@@ -23,294 +24,405 @@
...
   }
 
+  /// \name Data for compiled functions.
+  /// @{
+  std::vector<uint8_t *> MemoryPtrs;
+  std::vector<ValVariant *> GlobalPtrs;
+  /// @}
+
+  friend class Runtime::StoreManager;
+  using StoreCallback = void(StoreManager *Store, const ModuleInstance *Mod);
```
Comment (ibmibmibm) [inline include/runtime/instance/module.h:412]: `noexcept`

File: include/runtime/instance/module.h [hunk 2]
```diff
@@ -23,294 +24,405 @@
...
+
 class ModuleInstance {
 public:
   ModuleInstance(std::string_view Name) : ModName(Name) {}
-  ~ModuleInstance() = default;
+  virtual ~ModuleInstance() {
+    // When destroying this module instance, call the callbacks to unlink to the
+    // store managers.
+    for (auto &&Pair : LinkedStore) {
+      Pair.second(Pair.first, this);
```
Comment (ibmibmibm) [inline include/runtime/instance/module.h:72]: `assuming(Pair.second);`

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

File: include/runtime/storemgr.h [hunk 3]
```diff
@@ -13,333 +13,92 @@
...
+    std::unique_lock Lock(Mutex);
+    auto Iter = NamedMod.find(ModInst->getModuleName());
+    if (likely(Iter != NamedMod.cend())) {
+      return Unexpect(ErrCode::ModuleNameConflict);
+    }
+    NamedMod.emplace(ModInst->getModuleName(), ModInst);
+    // Link the module instance to this store manager.
+    (const_cast<Instance::ModuleInstance *>(ModInst))
+        ->linkStore(this, [](StoreManager *Store,
+                             const Instance::ModuleInstance *Inst) {
```
Comment (ibmibmibm) [inline include/runtime/storemgr.h:80]: `noexcept`

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

File: include/runtime/storemgr.h (hunk 3 above)
Comment (q82419) [reply include/runtime/storemgr.h]: added.

**Comment by apepkuss** [comment]
Hi @q82419

Here is a scenario that the current design of `Vm` could fail to cover. The scenario is 

1. Create a host function and a `Table` instance; then, store the reference to the host function at some index of the table instance.
2. Create a `ModuleInstance` instance, and add the table instance (created in step 1) into the module instance. 
3. Create a `Vm` instance, then register the module instance (created in step 2).

After the steps above, we can get the internal store of the vm instance, and then get the table instance via the returned store, and then get the function reference to the host function via the table instance, but how to call the host function via the function reference?

**Comment by apepkuss** [comment]
@q82419 Thanks for the reply! I'll append the explanation for the scenario in the API doc. Thanks again!

File: include/runtime/instance/module.h [hunk 4]
```diff
@@ -23,294 +24,407 @@
...
-    FuncTypes.emplace_back(FuncType);
+    unsafeAddHostInstance(
+        Name, OwnedFuncInsts, FuncInsts, ExpFuncs,
+        std::make_unique<Runtime::Instance::FunctionInstance>(std::move(Func)));
   }
-
-  /// Register module owns instances with address in Store.
-  void addFunc(FunctionInstance *Func) {
+  void addHostFunc(std::string_view Name,
+                   std::unique_ptr<Instance::FunctionInstance> &Func) {
```
Comment (ibmibmibm) [inline include/runtime/instance/module.h:91]: should use `&&`, and merge with next function.

File: lib/executor/executor.cpp [hunk 5]
```diff
@@ -9,11 +9,13 @@
...
-Expect<void> Executor::instantiateModule(Runtime::StoreManager &StoreMgr,
-                                         const AST::Module &Mod) {
-  InsMode = InstantiateMode::Instantiate;
-  if (auto Res = instantiate(StoreMgr, Mod, ""); !Res) {
+/// Instantiate a WASM Module. See "include/executor/executor.h".
+Expect<std::unique_ptr<Runtime::Instance::ModuleInstance>>
+Executor::instantiateModule(Runtime::StoreManager &StoreMgr,
+                            const AST::Module &Mod) {
+  if (auto Res = instantiate(StoreMgr, Mod)) {
+    return std::move(*Res);
```
Comment (ibmibmibm) [inline lib/executor/executor.cpp:17]: `return Res;`

File: lib/executor/executor.cpp [hunk 6]
```diff
@@ -22,51 +24,15 @@ Expect<void> Executor::instantiateModule(Runtime::StoreManager &StoreMgr,
...
-                                      std::string_view Name) {
-  InsMode = InstantiateMode::ImportWasm;
-  if (auto Res = instantiate(StoreMgr, Mod, Name); !Res) {
-    spdlog::error(ErrInfo::InfoRegistering(Name));
+/// Register a named WASM module. See "include/executor/executor.h".
+Expect<std::unique_ptr<Runtime::Instance::ModuleInstance>>
+Executor::registerModule(Runtime::StoreManager &StoreMgr,
+                         const AST::Module &Mod, std::string_view Name) {
+  if (auto Res = instantiate(StoreMgr, Mod, Name)) {
+    return std::move(*Res);
```
Comment (ibmibmibm) [inline lib/executor/executor.cpp:34]: `return Res;`

File: lib/executor/instantiate/global.cpp [hunk 7]
```diff
@@ -24,22 +23,19 @@ Expect<void> Executor::instantiate(Runtime::StoreManager &StoreMgr,
...
-    // Insert global instance to store manager.
-    Runtime::Instance::GlobalInstance *GlobInst = nullptr;
-    if (InsMode == InstantiateMode::Instantiate) {
-      GlobInst = StoreMgr.pushGlobal(GlobSeg.getGlobalType());
-    } else {
-      GlobInst = StoreMgr.importGlobal(GlobSeg.getGlobalType());
-    }
+    // Create and add the global instance into the module instance.
+    ModInst.addGlobal(GlobSeg.getGlobalType());
+    Runtime::Instance::GlobalInstance *GlobInst =
```
Comment (ibmibmibm) [inline lib/executor/instantiate/global.cpp:31]: `const auto Index = ModInst.getGlobalNum() - 1;`

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

File: lib/executor/instantiate/module.cpp [hunk 8]
```diff
@@ -50,39 +52,32 @@ Expect<void> Executor::instantiate(Runtime::StoreManager &StoreMgr,
 
   // Instantiate ImportSection and do import matching. (ImportSec)
   const AST::ImportSection &ImportSec = Mod.getImportSection();
-  if (auto Res = instantiate(StoreMgr, *ModInst, ImportSec); !Res) {
+  if (auto Res = instantiate(StoreMgr, *ModInst.get(), ImportSec); !Res) {
```
Comment (ibmibmibm) [inline lib/executor/instantiate/module.cpp:55]: why not `*ModInst`?

File: test/api/APIUnitTest.cpp [hunk 9]
```diff
@@ -193,16 +194,16 @@ WasmEdge_ImportObjectContext *createExternModule(std::string_view Name,
...
 
   WasmEdge_FunctionTypeDelete(HostFType);
-  return ImpObj;
+  return HostMod;
 }
 
 // Helper function to load wasm file into AST module.
-WasmEdge_ASTModuleContext *loadModule(const WasmEdge_ConfigureContext *Conf,
-                                      const char *Path) {
+WasmEdge_ASTModuleContext *loadModule
```
Comment (ibmibmibm) [inline test/api/APIUnitTest.cpp:205]: `[[maybe_unused]] WasmEdge_ASTModuleContext *loadModule(...)`

**Review by ibmibmibm — CHANGES_REQUESTED** [review]

File: test/api/APIUnitTest.cpp (hunk 9 above)
Comment (q82419) [reply test/api/APIUnitTest.cpp]: Removed.

File: lib/executor/instantiate/module.cpp (hunk 8 above)
Comment (q82419) [reply lib/executor/instantiate/module.cpp]: Fixed.

File: lib/executor/executor.cpp (hunk 6 above)
Comment (q82419) [reply lib/executor/executor.cpp:34]: Fixed.

File: lib/executor/instantiate/global.cpp (hunk 7 above)
Comment (q82419) [reply lib/executor/instantiate/global.cpp]: Updated.

**Review by ibmibmibm — APPROVED** [review]

**Pull request merged**